<a href="https://colab.research.google.com/github/arvindpanda2007/LLM_From_Scratch_solo/blob/tokenizer_v2/tokenization.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [5]:
!cd

In [6]:
from google.colab import files

uploaded = files.upload()

Saving the-verdict.txt to the-verdict.txt


In [7]:

with open("the-verdict.txt","r") as f:
    text_R = f.read()

In [8]:
print("Total number of character:", len(text_R))
print(text_R[:99])

Total number of character: 20479
I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no 


In [9]:
import re
text = "Hello, world. This, is a test."
result = re.split(r'(\s)', text)
print(result)
result = [item for item in result if item.strip() != ""]
print(result)

['Hello,', ' ', 'world.', ' ', 'This,', ' ', 'is', ' ', 'a', ' ', 'test.']
['Hello,', 'world.', 'This,', 'is', 'a', 'test.']


In [10]:
preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', text_R)
preprocessed = [item.strip() for item in preprocessed if item.strip()]
print(preprocessed)

['I', 'HAD', 'always', 'thought', 'Jack', 'Gisburn', 'rather', 'a', 'cheap', 'genius', '--', 'though', 'a', 'good', 'fellow', 'enough', '--', 'so', 'it', 'was', 'no', 'great', 'surprise', 'to', 'me', 'to', 'hear', 'that', ',', 'in', 'the', 'height', 'of', 'his', 'glory', ',', 'he', 'had', 'dropped', 'his', 'painting', ',', 'married', 'a', 'rich', 'widow', ',', 'and', 'established', 'himself', 'in', 'a', 'villa', 'on', 'the', 'Riviera', '.', '(', 'Though', 'I', 'rather', 'thought', 'it', 'would', 'have', 'been', 'Rome', 'or', 'Florence', '.', ')', '"', 'The', 'height', 'of', 'his', 'glory', '"', '--', 'that', 'was', 'what', 'the', 'women', 'called', 'it', '.', 'I', 'can', 'hear', 'Mrs', '.', 'Gideon', 'Thwing', '--', 'his', 'last', 'Chicago', 'sitter', '--', 'deploring', 'his', 'unaccountable', 'abdication', '.', '"', 'Of', 'course', 'it', "'", 's', 'going', 'to', 'send', 'the', 'value', 'of', 'my', 'picture', "'", 'way', 'up', ';', 'but', 'I', 'don', "'", 't', 'think', 'of', 'that', ',

In [11]:
all_words = sorted(set(preprocessed))
vocab_size = len(all_words)
print(vocab_size)

1130


In [12]:
vocab = {token:integer for integer,token in enumerate(all_words)}
for i, item in enumerate(vocab.items()):
    print(item)
    if i >= 50:
        break


('!', 0)
('"', 1)
("'", 2)
('(', 3)
(')', 4)
(',', 5)
('--', 6)
('.', 7)
(':', 8)
(';', 9)
('?', 10)
('A', 11)
('Ah', 12)
('Among', 13)
('And', 14)
('Are', 15)
('Arrt', 16)
('As', 17)
('At', 18)
('Be', 19)
('Begin', 20)
('Burlington', 21)
('But', 22)
('By', 23)
('Carlo', 24)
('Chicago', 25)
('Claude', 26)
('Come', 27)
('Croft', 28)
('Destroyed', 29)
('Devonshire', 30)
('Don', 31)
('Dubarry', 32)
('Emperors', 33)
('Florence', 34)
('For', 35)
('Gallery', 36)
('Gideon', 37)
('Gisburn', 38)
('Gisburns', 39)
('Grafton', 40)
('Greek', 41)
('Grindle', 42)
('Grindles', 43)
('HAD', 44)
('Had', 45)
('Hang', 46)
('Has', 47)
('He', 48)
('Her', 49)
('Hermia', 50)


In [13]:
from importlib.metadata import version
import tiktoken
print("tiktoken version:", version("tiktoken"))

tiktoken version: 0.13.0


In [14]:
tokenizer = tiktoken.get_encoding("gpt2")

text = (
    "Hello, do you like tea? <|endoftext|> In the sunlit terraces"
    "of someunknownPlace."
)
integers = tokenizer.encode(text, allowed_special={"<|endoftext|>"})
print(integers)

[15496, 11, 466, 345, 588, 8887, 30, 220, 50256, 554, 262, 4252, 18250, 8812, 2114, 1659, 617, 34680, 27271, 13]


In [15]:
print(len(text))
print(len(integers))

80
20


In [16]:
strings = tokenizer.decode(integers)
print(strings)

Hello, do you like tea? <|endoftext|> In the sunlit terracesof someunknownPlace.


BPE tokenizer or byte-pair encoding tokenizer is based upon a sophisticated tokenization scheme which was used to train llms such as ChatGPT

It utilizes the <| endoftext |> token to join two different segments of text

the BPE tokenizer used to train GPT-2,3 and the original model in ChatGPT had a vocabulary size of 50,257 with <| endoftext |> being assigned the largest token id of 50,256

It utilizes an algorithm which breaks down the unknown words not in its vocabulary into a sequence of smaller sub-words or even individual characters to encode it.

Example:


In [17]:
integers = tokenizer.encode("Kabjngvf :;wfwg")
print(integers)
strings = tokenizer.decode(integers)
print(strings)

[42, 397, 73, 782, 85, 69, 1058, 26, 86, 44482, 70]
Kabjngvf :;wfwg


As we see the "unknown" string was broken down by the tokenizer into a sequence of smaller segments which were effectively decoded too later on

In [18]:
import tiktoken
tokenizer = tiktoken.get_encoding("gpt2")
with open("the-verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()
enc_text = tokenizer.encode(raw_text)
print(len(enc_text))

5145


In [19]:
enc_sample = enc_text[50:]

context_size = 5
x = enc_sample[:context_size]
y = enc_sample[1:context_size+1]
print(f"x: {x}")
print(f"y:      {y}")

x: [290, 4920, 2241, 287, 257]
y:      [4920, 2241, 287, 257, 4489]


In [20]:
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]
    print(context, "---->", desired)

[290] ----> 4920
[290, 4920] ----> 2241
[290, 4920, 2241] ----> 287
[290, 4920, 2241, 287] ----> 257
[290, 4920, 2241, 287, 257] ----> 4489


In [21]:
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]
    print(tokenizer.decode(context), "---->", tokenizer.decode([desired]))

 and ---->  established
 and established ---->  himself
 and established himself ---->  in
 and established himself in ---->  a
 and established himself in a ---->  vill


In [22]:
from torch.utils.data import Dataset, DataLoader

In [23]:
import tiktoken
with open("the-verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()
tokenizer = tiktoken.get_encoding("gpt2")

In [24]:
import torch

In [25]:
class GPTDataset(Dataset):
    def __init__(self, text, tokenizer, context_size, stride):
        self.input_ids = []
        self.target_ids = []
        tokenized_text = tokenizer.encode(text)
        for i in range(0, len(tokenized_text) - context_size, stride):
            input_id = tokenized_text[i:i+context_size]
            target_id = tokenized_text[i+1:i+context_size+1]
            self.input_ids.append(torch.tensor(input_id))
            self.target_ids.append(torch.tensor(target_id))

    def __len__(self):
        return len(self.input_ids)

    def __getitem__(self, sample_id):
        return self.input_ids[sample_id], self.target_ids[sample_id]

In [26]:
dataset = GPTDataset(
    raw_text, tokenizer,
    context_size=4,
    stride=1
)

In [27]:
dataset[0]

(tensor([  40,  367, 2885, 1464]), tensor([ 367, 2885, 1464, 1807]))

In [28]:
len(dataset)

5141

In [29]:
raw_text[:100]

'I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no g'

In [30]:
def data_loader(text, batch_size, context_window = 256, stride = 64, shuffle = False,num_workers = 3,drop_last = True):
    tokenizer = tiktoken.get_encoding("gpt2")
    dataset = GPTDataset(
        text, tokenizer,
        context_size=context_window,
        stride=stride
    )
    return DataLoader(dataset, batch_size=batch_size, shuffle=shuffle, num_workers=num_workers, drop_last=drop_last)


In [31]:
dataloader = data_loader(raw_text, batch_size=4, context_window = 256, stride = 64, shuffle=True, num_workers=3)

/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py:424: UserWarning: This DataLoader will create 3 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  self.check_worker_number_rationality()


In [32]:
for batch_x, batch_y in dataloader:

    print(batch_x)

    print(batch_y)

    break

/usr/local/lib/python3.12/dist-packages/torch/utils/data/dataloader.py:432: UserWarning: This DataLoader will create 3 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  self.check_worker_number_rationality()


tensor([[  271, 10899,   550,  ..., 29543,  2745,    11],
        [ 7163,   262,  8631,  ...,  1654,   301,   835],
        [ 1310,  1165,   881,  ...,    13,   366,  2215],
        [ 1497,    11,   290,  ...,   270,  7997,    11]])
tensor([[10899,   550,   366,  ...,  2745,    11,   314],
        [  262,  8631, 26210,  ...,   301,   835,   286],
        [ 1165,   881, 40642,  ...,   366,  2215,   673],
        [   11,   290,   531,  ...,  7997,    11,   319]])


The next step after tokenization of raw text sample is to convert them into input-target pairs for training of an llm

for this we will implement two classes
1. Dataset class that inherits __init__ , __len__ , __getitem__ methods from Dataset class in Pytorch library
    a. __init__ method involves taking the text encodings list and initializing input_id list and a target_id list which would later be filled with tensors of inputs and targets respectively
    Note: The tensor arrays input_id and target_id is initially just two lists with tensor elements but due to torch internal mechanism of torch.stack() the entire thing is converted to a tensor

    b. __len__ returns the total number of elements in input_id or target_id or alternatively the total number of input-target pairs

    c. __getitem__ method takes the sample id that is the element number of input_id and returns the corresponding input - target pair

Now after generating the pairs we need to convert this data into batches of input(x) and target(y) stacked together
**Why?**
Because if each pair were dealt with individually then it would really be a lot of waste of computational power since GPUs are built for parallelism
For this we will use the DataLoader class
2. Dataloader class uses the Dataset class previously created and creates batches using Dataloader class in pytorch library


##IMPORTANT PARAMETERS

1. Batch Size: How many samples per batch

2. Context Window: How many tokens can the llm see before predicting next word

3. Shuffle: Shuffling in Dataloader shuffles the order of input and target in their respective batches. It is preferred since it prevents memorization of order

4. Drop_Last: Suppose there is a case where distribution of samples in each input- target pair is same but in the last pair there are very less samples left so distribution becomes lesser then drop last removes it

5. num_workers: num_workers is the number of separate CPU processes PyTorch uses to prepare data in parallel.

6. Stride: stride controls how far the sliding window moves when creating training sequences.

##How these Parameters Affect the Process

*   Larger Batch size improves GPU utilization and faster training but also increases memory usage.
*   Larger Context size improves the models reasoning capability and understanding longer contexts but also increases attention cost.
*   Smaller Stride means more training samples but also means more overlap.
*   Shuffling leads to better convergence and better prevents overfitting of patterns.
*   drop_last = true means there are fixed tensor shapes and stable GPU performance
*   more workers means faster batch preparation and better GPU utilization but increases overhead RAM usage.





##CREATING TOKEN EMBEDDINGS

The last step in preparing input text for LLM training is to convert token IDs to embedding vectors.

The first step is to initialize embedding weights with random values.
Then the model optimizes the vectors based on its discovery of percieved relationships  using backpropagation.

**Now what is backpropagation?**
During the training process the model makes predictions for the next word and then checks the actual result. If they do not match then the model goes back adjusts its internal weights and repeats the procedures. This is called backpropagation mechanism.

The code for the first step is as shown below ☕:


In [33]:
import torch

#Suppose we have 5 input tokens 2,3,5,1,18

input_token_ids = torch.tensor([2,3,5,1,18])
vocab_size = 19
output_dim = 3

torch.manual_seed(124)

embedding_layer = torch.nn.Embedding(vocab_size, output_dim)

print(embedding_layer)

Embedding(19, 3)


In [34]:
print(embedding_layer.weight)

Parameter containing:
tensor([[-0.5728,  0.2498,  1.2100],
        [-1.0713,  1.1983,  0.4496],
        [ 0.0454, -1.0395,  0.0552],
        [ 1.3901,  1.4865,  0.3712],
        [-0.0122, -0.2461,  0.8598],
        [-0.4195, -0.0496, -0.9707],
        [-1.2034, -1.5485, -0.2655],
        [ 0.8484,  0.8453, -1.4826],
        [-0.4124,  0.6751, -0.8403],
        [ 1.4388, -0.9206,  0.1130],
        [-0.7559, -0.3033, -1.3482],
        [ 0.7074, -0.8038, -0.2067],
        [ 0.3137,  1.7624, -0.2038],
        [-0.3140,  1.0573,  1.0477],
        [ 0.0862,  0.2267, -0.4979],
        [ 2.3254, -1.2368,  0.0045],
        [-1.7099,  0.5037, -0.1479],
        [ 0.1531, -0.1364, -0.3311],
        [ 1.5150, -0.9470, -1.3890]], requires_grad=True)


In [35]:
print(embedding_layer.weight.shape)

torch.Size([19, 3])


In [36]:
print(embedding_layer(input_token_ids))

tensor([[ 0.0454, -1.0395,  0.0552],
        [ 1.3901,  1.4865,  0.3712],
        [-0.4195, -0.0496, -0.9707],
        [-1.0713,  1.1983,  0.4496],
        [ 1.5150, -0.9470, -1.3890]], grad_fn=<EmbeddingBackward0>)


Now embedding_layer is a pytorch object that is not only having a tensor matrix of weights but also has methods and other parameters related to it such as **requires_grad** (which means that backpropagation is allowed on weights).

Now here's a problem

The embedding layer has converted each token ID from the sentence into embeddings but the transformer does not inherently see the sentence based on these embeddings

**Example** : Man bit Dog                                                          
             Dog bit Man

in both cases the same input token ids are put in and the same embeddings come but the position is not known by the transformer

So the solution is also the to additionally include a positional embedding which says its position.

In [37]:
import torch

vocab_size = 50257
output_dim = 256
embedding_layer = torch.nn.Embedding(vocab_size, output_dim)

In [38]:
from google.colab import files
uploaded = files.upload()

Saving the-verdict.txt to the-verdict (1).txt


In [39]:
with open(r"the-verdict.txt","r") as f:
  raw_text = f.read()

In [40]:
import tiktoken

tokenizer = tiktoken.get_encoding("gpt2")

In [41]:
from torch.utils.data import Dataset, DataLoader

In [42]:
class dataset(Dataset):
  def __init__(self, text, tokenizer, context_size, stride):
    self.input_ids = []
    self.target_ids = []
    tokenized_text = tokenizer.encode(text)
    for i in range(0, len(tokenized_text) - context_size, stride):
      input_id = tokenized_text[i:i+context_size]
      target_id = tokenized_text[i+1:i+context_size+1]
      self.input_ids.append(torch.tensor(input_id))
      self.target_ids.append(torch.tensor(target_id))
    return None

  def __len__(self):
    return len(self.input_ids)

  def __getitem__(self, sample_id):
    return self.input_ids[sample_id], self.target_ids[sample_id]


In [43]:
dataset_rawtext = dataset(raw_text,tokenizer,4,4)

In [44]:
dataset_rawtext.__getitem__(3)

(tensor([15632,   438,  2016,   257]), tensor([ 438, 2016,  257,  922]))

In [45]:
def dataloader(text, batch_size, context_window = 256, stride = 64, shuffle = False,num_workers = 2,drop_last = True):
    tokenizer = tiktoken.get_encoding("gpt2")
    dataset_text = dataset(
        text, tokenizer,
        context_size=context_window,
        stride=stride
    )
    return DataLoader(dataset_text, batch_size=batch_size, shuffle=shuffle, num_workers=num_workers, drop_last=drop_last)

In [46]:
dataloader_rawtext = dataloader(raw_text, 12, 256, 64, True, 2, True)

In [47]:
print(dataloader_rawtext.dataset[0])

(tensor([   40,   367,  2885,  1464,  1807,  3619,   402,   271, 10899,  2138,
          257,  7026, 15632,   438,  2016,   257,   922,  5891,  1576,   438,
          568,   340,   373,   645,  1049,  5975,   284,   502,   284,  3285,
          326,    11,   287,   262,  6001,   286,   465, 13476,    11,   339,
          550,  5710,   465, 12036,    11,  6405,   257,  5527, 27075,    11,
          290,  4920,  2241,   287,   257,  4489,    64,   319,   262, 34686,
        41976,    13,   357, 10915,   314,  2138,  1807,   340,   561,   423,
          587, 10598,   393, 28537,  2014,   198,   198,     1,   464,  6001,
          286,   465, 13476,     1,   438,  5562,   373,   644,   262,  1466,
         1444,   340,    13,   314,   460,  3285,  9074,    13, 46606,   536,
         5469,   438, 14363,   938,  4842,  1650,   353,   438,  2934,   489,
         3255,   465, 48422,   540,   450,    67,  3299,    13,   366,  5189,
         1781,   340,   338,  1016,   284,  3758,   262,  1988,

In [48]:
vocab_size = 50257
embedding_dim = 256

token_embedding = torch.nn.Embedding(
    vocab_size,
    embedding_dim
)

In [49]:
for inputs, targets in dataloader_rawtext:
  print(inputs.shape)
  break

torch.Size([12, 256])


In [50]:
embedding_inputs = token_embedding(inputs)

In [51]:
print(embedding_inputs.shape)

torch.Size([12, 256, 256])


In [52]:
context_window = 256
pos_embedding_layer = torch.nn.Embedding(
    vocab_size,
    embedding_dim
)

In [53]:
pos_embeddings = pos_embedding_layer(torch.arange(context_window))

In [54]:
print(pos_embeddings.shape)

torch.Size([256, 256])


In [55]:
input_embeddings = embedding_inputs + pos_embeddings

In [56]:
print(input_embeddings)

tensor([[[ 6.8902e-01, -6.9332e-02,  1.1071e+00,  ..., -1.7679e+00,
           7.4834e-01,  4.3034e-01],
         [ 1.1844e+00,  7.0539e-01, -1.2938e+00,  ...,  3.0485e-01,
           1.0396e+00, -2.9283e-01],
         [-4.1888e-02, -3.0519e+00,  7.7967e-01,  ..., -3.4706e-01,
          -5.3284e-01,  1.5125e+00],
         ...,
         [ 1.1195e+00,  3.1697e-02,  2.4078e-01,  ...,  1.1201e+00,
           1.9529e+00, -5.4922e-02],
         [ 7.7210e-01, -1.5333e+00,  1.0069e+00,  ..., -1.5978e+00,
          -2.9826e-01, -7.7733e-01],
         [ 4.6454e-03, -2.4931e-01,  5.6090e-01,  ..., -2.4937e+00,
          -1.2531e+00,  8.6972e-02]],

        [[ 1.2693e+00, -4.4109e-01,  7.5025e-01,  ...,  1.2520e+00,
          -3.3814e-01,  1.6156e+00],
         [-5.0881e-01, -3.9782e-01, -3.5207e-01,  ..., -2.9543e-01,
           8.0570e-01, -6.1873e-01],
         [ 1.3802e+00, -1.1548e+00,  1.7213e+00,  ...,  4.0807e-01,
          -1.1948e+00,  9.1708e-01],
         ...,
         [-3.1981e+00,  1

In [57]:
import torch
inputs = torch.tensor(
  [[0.43, 0.15, 0.89], # Your    (x^1)
   [0.55, 0.87, 0.66], # journey  (x^2)
   [0.57, 0.85, 0.64], # starts  (x^3)
   [0.22, 0.58, 0.33], # with   (x^4)
   [0.77, 0.25, 0.10], # one    (x^5)
   [0.05, 0.80, 0.55]] # step   (x^6)
)

In [58]:
query = inputs[1]
attn_scores_2 = torch.empty(inputs.shape[0])
for i, x_i in enumerate(inputs):
    attn_scores_2[i] = torch.dot(x_i, query)
print(attn_scores_2)

tensor([0.9544, 1.4950, 1.4754, 0.8434, 0.7070, 1.0865])


In [59]:
attn_weights_2_tmp = attn_scores_2 / attn_scores_2.sum()
print("Attention weights:", attn_weights_2_tmp)
print("Sum:", attn_weights_2_tmp.sum())

Attention weights: tensor([0.1455, 0.2278, 0.2249, 0.1285, 0.1077, 0.1656])
Sum: tensor(1.0000)


In [60]:
def softmax_naive(x):
    return torch.exp(x) / torch.exp(x).sum(dim=0)
attn_weights_2_naive = softmax_naive(attn_scores_2)
print("Attention weights:", attn_weights_2_naive)
print("Sum:", attn_weights_2_naive.sum())

Attention weights: tensor([0.1385, 0.2379, 0.2333, 0.1240, 0.1082, 0.1581])
Sum: tensor(1.)


In [61]:
#the softmax function is a better way to normalize attention weights since it gives higher values to important tokens and doesnt just distribute it to irrelevant info.

In [62]:
'''Note that this naive softmax implementation (softmax_naive) may encounter
numerical instability problems, such as overflow and underflow, when dealing with
large or small input values. Therefore, in practice, it’s advisable to use the PyTorch
implementation of softmax, which has been extensively optimized for performance:'''

'Note that this naive softmax implementation (softmax_naive) may encounter\nnumerical instability problems, such as overflow and underflow, when dealing with\nlarge or small input values. Therefore, in practice, it’s advisable to use the PyTorch\nimplementation of softmax, which has been extensively optimized for performance:'

In [63]:
attn_weights_2 = torch.softmax(attn_scores_2, dim=0)
print("Attention weights:", attn_weights_2)
print("Sum:", attn_weights_2.sum())

Attention weights: tensor([0.1385, 0.2379, 0.2333, 0.1240, 0.1082, 0.1581])
Sum: tensor(1.)


In [64]:
#This is the final representation of the context token
query = inputs[1]
context_vec_2 = torch.zeros(query.shape)
for i,x_i in enumerate(inputs):
    context_vec_2 += attn_weights_2[i]*x_i
print(context_vec_2)

tensor([0.4419, 0.6515, 0.5683])


In [65]:
attn_scores = torch.empty(6, 6)
for i, x_i in enumerate(inputs):
    for j, x_j in enumerate(inputs):
        attn_scores[i, j] = torch.dot(x_i, x_j)
print(attn_scores)

tensor([[0.9995, 0.9544, 0.9422, 0.4753, 0.4576, 0.6310],
        [0.9544, 1.4950, 1.4754, 0.8434, 0.7070, 1.0865],
        [0.9422, 1.4754, 1.4570, 0.8296, 0.7154, 1.0605],
        [0.4753, 0.8434, 0.8296, 0.4937, 0.3474, 0.6565],
        [0.4576, 0.7070, 0.7154, 0.3474, 0.6654, 0.2935],
        [0.6310, 1.0865, 1.0605, 0.6565, 0.2935, 0.9450]])


When computing the preceding attention score tensor, we used for loops in
Python. However, for loops are generally slow, and we can achieve the same results
using matrix multiplication:

In [66]:
attn_scores = inputs @ inputs.T
print(attn_scores)

tensor([[0.9995, 0.9544, 0.9422, 0.4753, 0.4576, 0.6310],
        [0.9544, 1.4950, 1.4754, 0.8434, 0.7070, 1.0865],
        [0.9422, 1.4754, 1.4570, 0.8296, 0.7154, 1.0605],
        [0.4753, 0.8434, 0.8296, 0.4937, 0.3474, 0.6565],
        [0.4576, 0.7070, 0.7154, 0.3474, 0.6654, 0.2935],
        [0.6310, 1.0865, 1.0605, 0.6565, 0.2935, 0.9450]])


In [67]:
all_context_vecs = attn_scores @ inputs
print(all_context_vecs)

tensor([[1.9802, 2.6760, 2.6721],
        [2.8579, 4.2330, 3.7270],
        [2.8335, 4.1718, 3.6734],
        [1.5501, 2.4086, 2.0693],
        [1.5969, 1.8945, 1.6743],
        [1.8911, 3.1515, 2.7231]])


In [68]:
x_2 = inputs[1]
d_in = inputs.shape[1]
d_out = 2

In [69]:
torch.manual_seed(123)
W_query = torch.nn.Parameter(torch.rand(d_in, d_out), requires_grad=False)
W_key   = torch.nn.Parameter(torch.rand(d_in, d_out), requires_grad=False)
W_value = torch.nn.Parameter(torch.rand(d_in, d_out), requires_grad=False)

In [70]:
query_2 = x_2 @ W_query
key_2 = x_2 @ W_key
value_2 = x_2 @ W_value
print(query_2)

tensor([0.4306, 1.4551])


In [71]:
keys = inputs @ W_key
values = inputs @ W_value
print("keys.shape:", keys.shape)
print("values.shape:", values.shape)

keys.shape: torch.Size([6, 2])
values.shape: torch.Size([6, 2])


In [72]:
keys.shape: torch.Size([6, 2])
values.shape: torch.Size([6, 2])

In [73]:
keys_2 = keys[1]
attn_score_22 = query_2.dot(keys_2)
print(attn_score_22)

tensor(1.8524)


In [74]:
attn_scores_2 = query_2 @ keys.T
print(attn_scores_2)

tensor([1.2705, 1.8524, 1.8111, 1.0795, 0.5577, 1.5440])


In [75]:
d_k = keys.shape[-1]
attn_weights_2 = torch.softmax(attn_scores_2 / d_k**0.5, dim=-1)
print(attn_weights_2)

tensor([0.1500, 0.2264, 0.2199, 0.1311, 0.0906, 0.1820])


In [76]:
context_vec_2 = attn_weights_2 @ values
print(context_vec_2)

tensor([0.3061, 0.8210])


##Why query, key, and value?
The terms “key,” “query,” and “value” in the context of attention mechanisms are
borrowed from the domain of information retrieval and databases, where similar con
cepts are used to store, search, and retrieve information.
A query is analogous to a search query in a database. It represents the current item
(e.g., a word or token in a sentence) the model focuses on or tries to understand.
The query is used to probe the other parts of the input sequence to determine how
much attention to pay to them.
The key is like a database key used for indexing and searching. In the attention mech
anism, each item in the input sequence (e.g., each word in a sentence) has an asso
ciated key. These keys are used to match the query.
The value in this context is similar to the value in a key-value pair in a database. It
represents the actual content or representation of the input items. Once the model
determines which keys (and thus which parts of the input) are most relevant to the
query (the current focus item), it retrieves the corresponding values.

##The rationale behind scaled-dot product attention
The reason for the normalization by the embedding dimension size is to improve the
training performance by avoiding small gradients. For instance, when scaling up the
embedding dimension, which is typically greater than 1,000 for GPT-like LLMs, large
dot products can result in very small gradients during backpropagation due to the
softmax function applied to them. As dot products increase, the softmax function
behaves more like a step function, resulting in gradients nearing zero. These small
gradients can drastically slow down learning or cause training to stagnate.
The scaling by the square root of the embedding dimension is the reason why this
self-attention mechanism is also called scaled-dot product attention.

In [77]:
import torch.nn as nn
class SelfAttention_v1(nn.Module):
    def __init__(self, d_in, d_out):
        super().__init__()
        self.W_query = nn.Parameter(torch.rand(d_in, d_out))
        self.W_key   = nn.Parameter(torch.rand(d_in, d_out))
        self.W_value = nn.Parameter(torch.rand(d_in, d_out))
    def forward(self, x):
        keys = x @ self.W_key
        queries = x @ self.W_query
        values = x @ self.W_value
        attn_scores = queries @ keys.T # omega
        attn_weights = torch.softmax(
            attn_scores / keys.shape[-1]**0.5, dim=-1
        )
        context_vec = attn_weights @ values
        return context_vec

In [78]:
import torch
torch.manual_seed(123)
sa_v1 = SelfAttention_v1(d_in, d_out)
print(sa_v1(inputs))

tensor([[0.2996, 0.8053],
        [0.3061, 0.8210],
        [0.3058, 0.8203],
        [0.2948, 0.7939],
        [0.2927, 0.7891],
        [0.2990, 0.8040]], grad_fn=<MmBackward0>)


We can improve the SelfAttention_v1 implementation further by utilizing
PyTorch’s nn.Linear layers, which effectively perform matrix multiplication when
the bias units are disabled. Additionally, a significant advantage of using nn.Linear instead of manually implementing nn.Parameter(torch.rand(...)) is that nn.Linear has an optimized weight initialization scheme, contributing to more stable and effective model training

In [79]:
class SelfAttention_v2(nn.Module):
    def __init__(self, d_in, d_out, qkv_bias=False):
        super().__init__()
        self.W_query = nn.Linear(d_in, d_out, bias=qkv_bias)
        self.W_key   = nn.Linear(d_in, d_out, bias=qkv_bias)
        self.W_value = nn.Linear(d_in, d_out, bias=qkv_bias)
    def forward(self, x):
        keys = self.W_key(x)
        queries = self.W_query(x)
        values = self.W_value(x)
        attn_scores = queries @ keys.T
        attn_weights = torch.softmax(
            attn_scores / keys.shape[-1]**0.5, dim=-1
        )
        context_vec = attn_weights @ values
        return context_vec

In [80]:
torch.manual_seed(789)
sa_v2 = SelfAttention_v2(d_in, d_out)
print(sa_v2(inputs))

tensor([[-0.0739,  0.0713],
        [-0.0748,  0.0703],
        [-0.0749,  0.0702],
        [-0.0760,  0.0685],
        [-0.0763,  0.0679],
        [-0.0754,  0.0693]], grad_fn=<MmBackward0>)


They give different weights since both follow different weight initialization schemes and the latter has a more sophisticated version of all this

Imagine reading a book with 8 specialists sitting beside you:

One tracks characters.
One tracks locations.
One tracks emotions.
One tracks cause and effect.
One tracks grammar.
One tracks chronology.

They all read only the pages that have already been written (causal mask), but each focuses on different patterns (multi-head attention). At the end, they combine their notes to make a prediction about the next word.

That's essentially multi-head causal self-attention in a Transformer.

In [81]:
queries = sa_v2.W_query(inputs)
keys = sa_v2.W_key(inputs)
attn_scores = queries @ keys.T
attn_weights = torch.softmax(attn_scores / keys.shape[-1]**0.5, dim=-1)
print(attn_weights)

tensor([[0.1921, 0.1646, 0.1652, 0.1550, 0.1721, 0.1510],
        [0.2041, 0.1659, 0.1662, 0.1496, 0.1665, 0.1477],
        [0.2036, 0.1659, 0.1662, 0.1498, 0.1664, 0.1480],
        [0.1869, 0.1667, 0.1668, 0.1571, 0.1661, 0.1564],
        [0.1830, 0.1669, 0.1670, 0.1588, 0.1658, 0.1585],
        [0.1935, 0.1663, 0.1666, 0.1542, 0.1666, 0.1529]],
       grad_fn=<SoftmaxBackward0>)


In [82]:
context_length = attn_scores.shape[0]
mask_simple = torch.tril(torch.ones(context_length, context_length))
print(mask_simple)


tensor([[1., 0., 0., 0., 0., 0.],
        [1., 1., 0., 0., 0., 0.],
        [1., 1., 1., 0., 0., 0.],
        [1., 1., 1., 1., 0., 0.],
        [1., 1., 1., 1., 1., 0.],
        [1., 1., 1., 1., 1., 1.]])


In [83]:
masked_simple = attn_weights*mask_simple
print(masked_simple)

tensor([[0.1921, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.2041, 0.1659, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.2036, 0.1659, 0.1662, 0.0000, 0.0000, 0.0000],
        [0.1869, 0.1667, 0.1668, 0.1571, 0.0000, 0.0000],
        [0.1830, 0.1669, 0.1670, 0.1588, 0.1658, 0.0000],
        [0.1935, 0.1663, 0.1666, 0.1542, 0.1666, 0.1529]],
       grad_fn=<MulBackward0>)


In [84]:
row_sums = masked_simple.sum(dim=-1, keepdim=True)
masked_simple_norm = masked_simple / row_sums
print(masked_simple_norm)

tensor([[1.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.5517, 0.4483, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.3800, 0.3097, 0.3103, 0.0000, 0.0000, 0.0000],
        [0.2758, 0.2460, 0.2462, 0.2319, 0.0000, 0.0000],
        [0.2175, 0.1983, 0.1984, 0.1888, 0.1971, 0.0000],
        [0.1935, 0.1663, 0.1666, 0.1542, 0.1666, 0.1529]],
       grad_fn=<DivBackward0>)


In [85]:
mask = torch.triu(torch.ones(context_length, context_length), diagonal=1)
masked = attn_scores.masked_fill(mask.bool(), -torch.inf)
print(masked)

tensor([[0.2899,   -inf,   -inf,   -inf,   -inf,   -inf],
        [0.4656, 0.1723,   -inf,   -inf,   -inf,   -inf],
        [0.4594, 0.1703, 0.1731,   -inf,   -inf,   -inf],
        [0.2642, 0.1024, 0.1036, 0.0186,   -inf,   -inf],
        [0.2183, 0.0874, 0.0882, 0.0177, 0.0786,   -inf],
        [0.3408, 0.1270, 0.1290, 0.0198, 0.1290, 0.0078]],
       grad_fn=<MaskedFillBackward0>)


In [86]:
attn_weights = torch.softmax(masked / keys.shape[-1]**0.5, dim=1)
print(attn_weights)

tensor([[1.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.5517, 0.4483, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.3800, 0.3097, 0.3103, 0.0000, 0.0000, 0.0000],
        [0.2758, 0.2460, 0.2462, 0.2319, 0.0000, 0.0000],
        [0.2175, 0.1983, 0.1984, 0.1888, 0.1971, 0.0000],
        [0.1935, 0.1663, 0.1666, 0.1542, 0.1666, 0.1529]],
       grad_fn=<SoftmaxBackward0>)


In [87]:
torch.manual_seed(123)
dropout = torch.nn.Dropout(0.5)
example = torch.ones(6, 6)
print(dropout(example))

tensor([[2., 2., 0., 2., 2., 0.],
        [0., 0., 0., 2., 0., 2.],
        [2., 2., 2., 2., 0., 2.],
        [0., 2., 2., 0., 0., 2.],
        [0., 2., 0., 2., 0., 2.],
        [0., 2., 2., 2., 2., 0.]])


In [88]:
torch.manual_seed(123)
print(dropout(attn_weights))

tensor([[2.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.7599, 0.6194, 0.6206, 0.0000, 0.0000, 0.0000],
        [0.0000, 0.4921, 0.4925, 0.0000, 0.0000, 0.0000],
        [0.0000, 0.3966, 0.0000, 0.3775, 0.0000, 0.0000],
        [0.0000, 0.3327, 0.3331, 0.3084, 0.3331, 0.0000]],
       grad_fn=<MulBackward0>)


In [89]:
batch = torch.stack((inputs, inputs), dim=0)
print(batch.shape)

torch.Size([2, 6, 3])


In [93]:
class CausalAttention(nn.Module):
    def __init__(self, d_in, d_out, context_length,
        dropout, qkv_bias=False):
        super().__init__()
        self.d_out = d_out
        self.W_query = nn.Linear(d_in, d_out, bias=qkv_bias)
        self.W_key   = nn.Linear(d_in, d_out, bias=qkv_bias)
        self.W_value = nn.Linear(d_in, d_out, bias=qkv_bias)
        self.dropout = nn.Dropout(dropout)
        self.register_buffer((
          'mask'),
        torch.triu(torch.ones(context_length, context_length),
        diagonal=1))
    def forward(self, x):
        b, num_tokens, d_in = x.shape
        keys = self.W_key(x)
        queries = self.W_query(x)
        values = self.W_value(x)
        attn_scores = queries @ keys.transpose(1, 2)
        attn_scores.masked_fill_(
            self.mask.bool()[:num_tokens, :num_tokens], -torch.inf)
        attn_weights = torch.softmax(
            attn_scores / keys.shape[-1]**0.5, dim=-1
        )
        attn_weights = self.dropout(attn_weights)
        context_vec = attn_weights @ values
        return context_vec

In [94]:
torch.manual_seed(123)
context_length = batch.shape[1]
ca = CausalAttention(d_in, d_out, context_length, 0.0)
context_vecs = ca(batch)
print("context_vecs.shape:", context_vecs.shape)

context_vecs.shape: torch.Size([2, 6, 2])
